src/internal/storage/fileset/fileset_test.go

```diff
@@ -1,274 +1,224 @@
-//		copy(wsCopy, ws)
-//		// Randomly distribute tagged data among filesets.
-//		for len(f.tags) > 0 {
-//			// Randomly select fileset to write to.
-//			i := rand.Intn(len(wsCopy))
-//			w := wsCopy[i]
-//			wsCopy = append(wsCopy[:i], wsCopy[i+1:]...)
-//			// Write the rest of the file if this is the last fileset.
-//			if len(wsCopy) == 0 {
-//				writeFile(t, w, &f, msg)
-//				break
-//			}
-//			// Choose a random number of the tags left.
-//			numTags := rand.Intn(len(f.tags)) + 1
-//			var size int
-//			for _, tag := range f.tags[:numTags] {
-//				size += int(tag.SizeBytes)
-//			}
-//			// Create file for writing and remove data/tags from rest of the file.
-//			fWrite := f
-//			fWrite.data = fWrite.data[:size]
-//			fWrite.tags = fWrite.tags[:numTags]
-//			f.data = f.data[size:]
-//			f.tags = f.tags[numTags:]
-//			writeFile(t, w, &fWrite, msg)
-//		}
-//	}
-//	for _, w := range ws {
-//		require.NoError(t, w.Close(), msg)
-//	}
-//	return files
-//}
-//
-//func getHashes(t *testing.T, fileSets *Storage, files []*testFile, msg string) {
-//	writeFileSet(t, fileSets, path.Join(scratchPath, Compacted), files, msg)
-//	r := fileSets.newReader(context.Background(), path.Join(scratchPath, Compacted))
-//	require.NoError(t, r.iterate(func(fr *FileReader) error {
-//		files[0].hashes = dataRefsToHashes(fr.Index().DataOp.DataRefs)
-//		files = files[1:]
-//		return nil
-//	}), msg)
-//}
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"testing"
+
+	units "github.com/docker/go-units"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/dbutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/require"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/track"
+)
+
+const (
+	max         = 20 * units.MB
+	maxTags     = 10
+	testPath    = "test"
+	scratchPath = "scratch"
+)
+
+type testFile struct {
+	path  string
+	parts []*testPart
+}
+
+type testPart struct {
+	tag  string
+	data []byte
+}
+
+func appendFile(t *testing.T, w *Writer, path string, parts []*testPart) {
+	// Write content and tags.
+	err := w.Append(path, func(fw *FileWriter) error {
+		for _, part := range parts {
+			fw.Append(part.tag)
+			if _, err := fw.Write(part.data); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	require.NoError(t, err)
+}
+
+func writeFileSet(t *testing.T, s *Storage, files []*testFile) ID {
+	w := s.NewWriter(context.Background())
+	for _, file := range files {
+		appendFile(t, w, file.path, file.parts)
+	}
+	id, err := w.Close()
+	require.NoError(t, err)
+	return *id
+}
+
+func checkFile(t *testing.T, f File, tf *testFile) {
+	r, w := io.Pipe()
+	eg := errgroup.Group{}
+	eg.Go(func() error {
+		return f.Content(w)
+	})
+	eg.Go(func() (retErr error) {
+		defer func() {
+			if retErr != nil {
+				r.CloseWithError(retErr)
+			} else {
+				r.Close()
```

Comment:
Is this close necessary?

Reply:
Probably not in the test, I think your are supposed to close it so that any writers will get errors instead of hanging.